Ensure images downloaded by pullTagV2 are always cleaned up

aarondav opened the pull request:
Previously, if only some of the downloads succeed, we would not close and delete the file handles. This does change the behavior of the registration to block on all downloads completing (whereas previously it would exit early after the first failure), which is necessary to ensure cleanup is only performed after the download portion has completed.

(We could also refactor the download function to take a signal when the registration has completed so it could wait and then clean itself up, but this would be a bit more involved.)

fixes https://github.com/docker/docker/issues/14506

aarondav commented on graph/pull.go:
```diff
@@ -538,14 +538,14 @@ func (s *TagStore) pullV2Tag(r *registry.Session, out io.Writer, endpoint *regis
 	out.Write(sf.FormatStatus(tag, "Pulling from %s", repoInfo.CanonicalName))
 
 	// downloadInfo is used to pass information from download to extractor
+	// tmpFile will be set if and only if the download was successful
```
previously, we checked both `downloaded` and `tmpFile`, but not quite correctly, so I just simplified by getting rid of `downloaded` since it provides no extra information, just opportunity for mistakes.

aarondav commented on graph/pull.go:
```diff
@@ -641,20 +640,27 @@ func (s *TagStore) pullV2Tag(r *registry.Session, out io.Writer, endpoint *regis
-		if d.err != nil {
-			if err := <-d.err; err != nil {
-				return false, err
+		registerFunc := func(d *downloadInfo, pullAborted bool) error {
```
I converted this into a function for three reasons:
1. to short-circuit the "pullAborted" case from trying to register
2. to invoke the deferred functions more quickly
3. in preparation for #14240 (i.e., fixing the FIXME below), where this will simplify the cleanup logic.

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "pull-cleanup" git@github.com:aarondav/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~2
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

thaJeztah commented:
Thanks @aarondav 

ping @stevvooe @dmcgowan @dmp42 for distribution

dmcgowan commented on graph/pull.go:
```diff
@@ -665,21 +676,27 @@ func (s *TagStore) pullV2Tag(r *registry.Session, out io.Writer, endpoint *regis
 		}
+		err := registerFunc(&downloads[i], lastErr != nil)
+		if err != nil {
+			lastErr = err
```
This is potentially returning a different error than before. Before the first error returned was the returned error, now it will be the last error.

dmcgowan commented on graph/pull.go:
```diff
@@ -641,20 +640,32 @@ func (s *TagStore) pullV2Tag(r *registry.Session, out io.Writer, endpoint *regis
+			}()
+
+			if d.err != nil {
+				if err := <-d.err; err != nil {
```
If there was prior error, this is going to wait for each of the other operations to finish. Why not check pullAborted before this, would also ensure the same error is returned.

aarondav replied:
This one was intentional, in order to cleanup the remaining downloads. This will be more critical for #14240.

dmcgowan replied:
Cleaning up the downloads is fine, but we need to return an error back to the client. The error returned should also be the first error not the last error. The first error is likely the most relevant.

aarondav replied:
Definitely agree. Done.

dmcgowan commented on graph/pull.go:
```diff
@@ -641,20 +640,32 @@ func (s *TagStore) pullV2Tag(r *registry.Session, out io.Writer, endpoint *regis
-			if err := <-d.err; err != nil {
-				return false, err
+		registerFunc := func(d *downloadInfo, pullAborted bool) error {
+			// Cleanup is guranteed to be invoked only after the download routine exits.
```
typo

aarondav replied:
Done.

dmcgowan commented:
I see what you trying to do here, setting to code review.

aarondav commented:
All outstanding comments addressed. Thanks for the review!

dmcgowan commented:
@aarondav if you addressed the comments can you force push the changes to review.

aarondav commented:
Ouch, my bad. Pushed.

aarondav commented:
Looks like a build break unrelated to my change.

dmcgowan commented:
@aarondav this is going to require a tricky rebase since thiis code was moved to `pull_v2.go`

aarondav committed:
Ensure images downloaded by pullTagV2 are always cleaned up

Previously, if only some of the downloads succeed, we would not close and delete the file handles. This does change the behavior of the registration to block on all downloads completing (whereas previously it would exit early after the first failure), which is necessary to ensure cleanup is only performed after the download portion has completed.

(We could also refactor the download function to take a signal when the registration has completed so it could wait and then clean itself up, but this would be a bit more involved.)

Signed-off-by: Aaron Davidson <aaron@databricks.com>

aarondav commented:
Rebased.

RichardScothern commented on graph/pull_v2.go:
```diff
@@ -244,22 +244,32 @@ func (p *v2Puller) pullV2Tag(tag, taggedName string) (bool, error) {
-		if d.err != nil {
-			if err := <-d.err; err != nil {
-				return false, err
+		registerFunc := func(d *downloadInfo, pullAborted bool) error {
```
Can you move this function definition out of the loop? It doesn't close over any variables and makes the logic harder to follow

tiborvass replied:
`out` and `p` are two variables that would also need to be passed.

@aarondav i suggest you make it a private method on `p` instead.

tiborvass commented:
@aarondav sorry but needs a rebase :(

calavera commented:
I'm carrying this in. Check #15414.

aarondav commented:
Thanks!

thaJeztah commented:
Thank YOU @aarondav ! :-)
